File: ssm-kim/boj/BOJ_2493.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] topHeight = new int[n];
        for (int i = 0; i < n; i++) {
            topHeight[i] = Integer.parseInt(st.nextToken());
        }

        ArrayDeque<Integer> stack =  new ArrayDeque<>();
        int[] answer = new int[n];

        // 뒤에서부터 탐색 (오른쪽에서 왼쪽으로)
        for (int i = n - 1; i >= 0; i--) {
            int current = topHeight[i];

            // 현재 탑보다 낮은 탑들은 현재 탑에서 수신 가능
            while (!stack.isEmpty() && current > topHeight[stack.peek()]) {
                int lastIdx = stack.pop();
                answer[lastIdx] = i + 1;
            }
            stack.push(i);  // 현재 탑을 스택에 추가 (다음 탐색 대상)
        }

        StringBuilder sb = new StringBuilder();
        for (int num : answer) {
            sb.append(num).append(" ");
        }
        System.out.println(sb);
    }
}

File: sodock00/boj/BOJ_2493.py
import sys
input = sys.stdin.readline

N = int(input())
tops = list(map(int, input().split(" ")))

# N : 1이상 500000이하
# 높이 : 1이상 100000000이하

# 각각의 탑에서 발사한 레이저 신호를 어느 탑에서 수신하는지 출력
# 수신하는 탑 없으면 0
# 나보다 크거나 같고, 가장 거리가 가까운 탑

# 그냥 반복문으로 돌리면 -> 시간초과
# 자료 구조 스택 사용해야함!

answer = [0] * N
stack = []
for i in range(N):
    while stack:
        if stack[-1][1] > tops[i]:
            answer[i] = stack[-1][0] + 1
            break
        else:
            stack.pop()
    stack.append((i, tops[i]))

print(answer)